queue/example_connection_pool_test.go

```diff
@@ -0,0 +1,148 @@
+package queue_test
+
+import (
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/tarantool/go-tarantool"
+	"github.com/tarantool/go-tarantool/connection_pool"
+	"github.com/tarantool/go-tarantool/queue"
+)
+
+type QueueConnectionHandler struct {
+	name string
+	cfg  queue.Cfg
+
+	uuid          uuid.UUID
+	registered    bool
+	err           error
+	registerMutex sync.Mutex
+	done          chan struct{}
+}
+
+var _ connection_pool.ConnectionHandler = &QueueConnectionHandler{}
+
+func NewQueueConnectionHandler(name string, cfg queue.Cfg) *QueueConnectionHandler {
+	return &QueueConnectionHandler{
+		name: name,
+		cfg:  cfg,
+		done: make(chan struct{}),
+	}
+}
+
+func (h *QueueConnectionHandler) Discovered(conn *tarantool.Connection,
+	role connection_pool.Role) error {
+	// The queue only works with a master instance.
+	if role != connection_pool.MasterRole {
+		return nil
+	}
+
+	// We need to register a shared session only once.
+	h.registerMutex.Lock()
+
+	if h.err != nil {
+		h.registerMutex.Unlock()
+		return h.err
+	}
+
+	q := queue.New(conn, h.name)
+	if !h.registered {
+		opts := queue.CfgOpts{InReplicaset: true, Ttr: 60 * time.Second}
+		if h.err = q.Cfg(opts); h.err == nil {
+			if h.uuid, h.err = q.Identify(nil); h.err == nil {
+				h.err = q.Create(h.cfg)
+			}
+		}
+		h.registered = h.err == nil
+		close(h.done)
+		h.registerMutex.Unlock()
+		return h.err
+	} else {
+		// We already registered. We don't need the lock any more.
+		h.registerMutex.Unlock()
```

Comment:
Isn't it all equivalent to `defer h.registerMutex.Unlock()`?

Reply:
No, the code unlocks the mutex immediately. Defer at the end of a function.

The idea here: we don't need to hold the lock in re-identify case because we already have a valid session UUID.

Reply:
Are concurrent `Identifiy`es and `Create`s valid?

Reply:
No, they are not.

We check `h.registered` value under the lock, it may become `true` only under the lock too. So if h.registered == true then `Create` already done.

Reply:
No, they're under the lock
```go
h.registerMutex.Unlock()
if _, err := q.Identify(&h.uuid); err != nil {
        return err
}
if err := q.Create(h.cfg); err != nil {
        return err
}
```

Reply:
So it still sounds like that unlock is dangerous.